### ISSUE TITLE
Using back button after opening and closing a dialog closes the app

### ISSUE THREAD

**Issue by myxzlpltk** [flutter/flutter#146341]
### Steps to reproduce

1. Open app (default page A)
2. Go to a page B inside `ShellRoute`
3. Go deeper into page C still inside `ShellRoute`
4. Open a dialog using `showDialog()`
5. Pop that dialog using `Navigator.of(context).pop()`
6. Press hardware back button

### Expected results

Pop back to page B

### Actual results

App closed

### Code sample

Clone this repository [https://github.com/myxzlpltk/routing_predictive_back](https://github.com/myxzlpltk/routing_predictive_back)

Hint: This behavior gone if you're not using `ShellRoute` or remove `android:enableOnBackInvokedCallback="true"` or setting it to false

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[weird behavior.webm](https://github.com/flutter/flutter/assets/59659668/dd4aef76-af0b-4262-bf52-18844b34a767)

</details>

### Logs

<details open><summary>Log during reproduce step:</summary>

```console
[+12790 ms] D/EGL_emulation(13580): app_time_stats: avg=5745.04ms min=156.12ms max=11333.95ms count=2
[  +78 ms] W/WindowOnBackDispatcher(13580): sendCancelIfRunning: isInProgress=falsecallback=io.flutter.embedding.android.FlutterActivity$1@7497194
[+1204 ms] D/EGL_emulation(13580): app_time_stats: avg=65.87ms min=1.60ms max=920.12ms count=17
[+1783 ms] D/EGL_emulation(13580): app_time_stats: avg=46.88ms min=9.56ms max=1101.39ms count=38
[+1430 ms] D/EGL_emulation(13580): app_time_stats: avg=35.23ms min=1.73ms max=785.75ms count=36
[ +417 ms] W/WindowOnBackDispatcher(13580): sendCancelIfRunning: isInProgress=falsecallback=io.flutter.embedding.android.FlutterActivity$1@7497194
[+4231 ms] D/EGL_emulation(13580): app_time_stats: avg=172.12ms min=12.94ms max=4217.97ms count=27
[ +673 ms] W/WindowOnBackDispatcher(13580): sendCancelIfRunning: isInProgress=falsecallback=android.app.Activity$$ExternalSyntheticLambda0@d9dec88
[+4329 ms] W/predictive_back(13580): Cleared Reference was only reachable from finalizer (only reported once)
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[!] Flutter (Channel stable, 3.19.5, on Microsoft Windows [Version 10.0.22621.3296], locale en-ID)
    • Flutter version 3.19.5 on channel stable at C:\Users\myxzlpltk\fvm\default
    ! Warning: `flutter` on your path resolves to C:\Users\myxzlpltk\fvm\versions\3.19.5\bin\flutter, which is not inside your current Flutter SDK checkout at
      C:\Users\myxzlpltk\fvm\default. Consider adding C:\Users\myxzlpltk\fvm\default\bin to the front of your path.
    ! Warning: `dart` on your path resolves to C:\Users\myxzlpltk\fvm\versions\3.19.5\bin\dart, which is not inside your current Flutter SDK checkout at
      C:\Users\myxzlpltk\fvm\default. Consider adding C:\Users\myxzlpltk\fvm\default\bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 300451adae (8 days ago), 2024-03-27 21:54:07 -0500
    • Engine revision e76c956498
    • Dart version 3.3.3
    • DevTools version 2.31.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at C:\Users\myxzlpltk\AppData\Local\Android\sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: C:\Users\myxzlpltk\AppData\Local\Programs\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0--11185874)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.9.0)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.9.34607.119
    • Windows 10 SDK version 10.0.22621.0

[√] Android Studio (version 2023.2)
    • Android Studio at C:\Users\myxzlpltk\AppData\Local\Programs\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.9+0--11185874)

[√] IntelliJ IDEA Community Edition (version 2023.3)
    • IntelliJ at C:\Users\myxzlpltk\AppData\Local\Programs\IntelliJ IDEA Community Edition
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart

[√] VS Code (version 1.87.2)
    • VS Code at C:\Users\myxzlpltk\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.86.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22621.3296]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 123.0.6312.106
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 123.0.2420.81

[√] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

**Comment by darshankawar** [comment]
@myxzlpltk 
Check below related / similar issues and see if your case resembles any of them or not:

https://github.com/flutter/flutter/issues/132504
https://github.com/flutter/flutter/issues/145290
https://github.com/flutter/flutter/issues/117061

**Comment by luskan** [comment]
Maybe it will be of use to somebody. 

I had similar problem. I have android native application in which we embed flutter fragment. We had a case, when user opened few flutter pages, and on the last one opened AlertDialog with edit text, used keyboard to enter text, then clicked mobile device back button. First click closed keyboard, the second one closed application. Instead of closing application AlertDialog should be closed.

We were creating embedded fragment like that:

flutterFragment = EmbeddedFlutterFragment.createFragment(engineCacheKey)

the solution was to set shouldAutomaticallyHandleOnBackPressed(true) like that:

      val builder = FlutterFragment.CachedEngineFragmentBuilder(EmbeddedFlutterFragment::class.java, engineCacheKey)
      flutterFragment = with(builder) {
        transparencyMode(TransparencyMode.transparent)
        destroyEngineWithFragment(false)
        shouldAutomaticallyHandleOnBackPressed(true)
        build()
      }

we are still in a process of testing, probably this is needed only for api 33 and 34. 

I am not sure if this is relevant here, but additionally I had to use WillPopScope at the bottom of widget tree, to prevent last back from closing android app, we have it handled in android activity (double back causes app exit)

My flutter : 3.19.5
